File: python/text/regex_query.py
"""Regex Query Tool — pattern matching, extraction, and replacement."""

import re
from typing import NamedTuple


class Match(NamedTuple):
    """A match with its position in the source text."""
    text: str
    start: int
    end: int


# Pre-built patterns for common structured data.
PATTERNS = {
    "EMAIL": r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}",
    "URL": r"https?://[^\s<>\"']+",
    "PHONE": r"\+?1?[-.\s]?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}",
    "DATE": r"\d{4}-\d{2}-\d{2}",
    "IP_ADDRESS": r"\b(?:\d{1,3}\.){3}\d{1,3}\b",
}


class RegexTool:
    """Apply regex patterns to text — find, extract, replace, split, validate."""

    patterns = PATTERNS

    def find_matches(self, text: str, pattern: str) -> list[str]:
        """Return all substrings matching *pattern*."""
        return re.findall(pattern, text)

    def find_with_positions(self, text: str, pattern: str) -> list[Match]:
        """Return matches with their (start, end) positions."""
        return [Match(m.group(), m.start(), m.end()) for m in re.finditer(pattern, text)]

    def replace(self, text: str, pattern: str, replacement: str) -> str:
        """Replace all occurrences of *pattern* with *replacement*."""
        return re.sub(pattern, replacement, text)

    def split(self, text: str, pattern: str) -> list[str]:
        """Split *text* on every occurrence of *pattern*."""
        return re.split(pattern, text)

    def is_match(self, text: str, pattern: str) -> bool:
        """Return True if *pattern* matches anywhere in *text*."""
        return re.search(pattern, text) is not None

    def extract_groups(self, text: str, pattern: str) -> list[tuple[str, ...]]:
        """Return capture-group tuples for every match of *pattern*."""
        return re.findall(pattern, text)


if __name__ == "__main__":
    tool = RegexTool()

    sample = (
        "Contact alice@example.com or bob@work.org. "
        "Visit https://example.com and https://docs.python.org/3 for info. "
        "Event on 2026-04-16, backup on 2026-05-01. "
        "Call 555-867-5309 or (212) 555-1234. "
        "Server at 192.168.1.1 and 10.0.0.1."
    )

    print("=== Find emails ===")
    emails = tool.find_matches(sample, PATTERNS["EMAIL"])
    for e in emails:
        print(f"  {e}")

    print("\n=== Extract URLs with positions ===")
    urls = tool.find_with_positions(sample, PATTERNS["URL"])
    for m in urls:
        print(f"  {m.text}  [{m.start}:{m.end}]")

    print("\n=== Replace dates (YYYY-MM-DD -> DD/MM/YYYY) ===")
    replaced = tool.replace(sample, r"(\d{4})-(\d{2})-(\d{2})", r"\3/\2/\1")
    print(f"  {replaced}")

    print("\n=== Split on commas ===")
    csv_line = "apples, bananas, cherries, dates"
    parts = tool.split(csv_line, r",\s*")
    print(f"  {parts}")

    print("\n=== Validate IP addresses ===")
    tests = ["192.168.1.1", "not-an-ip", "10.0.0.1", "999.999.999.999"]
    for t in tests:
        valid = tool.is_match(t, r"^" + PATTERNS["IP_ADDRESS"] + r"$")
        print(f"  {t}: {valid}")

    print("\n=== Extract groups (parse log entries) ===")
    log = "[2026-04-16 08:30:00] ERROR: disk full\n[2026-04-16 09:15:22] INFO: recovered"
    entries = tool.extract_groups(log, r"\[(\d{4}-\d{2}-\d{2}) (\d{2}:\d{2}:\d{2})\] (\w+): (.+)")
    for date, time, level, msg in entries:
        print(f"  {date} {time} [{level}] {msg}")

File: rust/text/regex_query.rs
//! Regex Query Tool — pattern matching, extraction, and replacement.
//!
//! Since the `regex` crate is an external dependency, this module provides
//! pre-built pattern matchers for common structured data (email, URL, phone,
//! date, IP address) using character-level scanning, plus generic helpers
//! that delegate to `regex::Regex` when the crate is available.

/// A match result carrying the matched text and its byte-offset span.
#[derive(Debug, Clone, PartialEq)]
pub struct Match {
    pub text: String,
    pub start: usize,
    pub end: usize,
}

// ---------------------------------------------------------------------------
// Generic regex operations (require the `regex` crate)
// ---------------------------------------------------------------------------

/// Find all non-overlapping matches of `pattern` in `text`.
pub fn find_matches(text: &str, pattern: &str) -> Vec<String> {
    let re = regex::Regex::new(pattern).expect("invalid regex");
    re.find_iter(text).map(|m| m.as_str().to_string()).collect()
}

/// Find all matches with their byte-offset positions.
pub fn find_with_positions(text: &str, pattern: &str) -> Vec<Match> {
    let re = regex::Regex::new(pattern).expect("invalid regex");
    re.find_iter(text)
        .map(|m| Match {
            text: m.as_str().to_string(),
            start: m.start(),
            end: m.end(),
        })
        .collect()
}

/// Replace all occurrences of `pattern` with `replacement`.
pub fn replace(text: &str, pattern: &str, replacement: &str) -> String {
    let re = regex::Regex::new(pattern).expect("invalid regex");
    re.replace_all(text, replacement).into_owned()
}

/// Split `text` on every occurrence of `pattern`.
pub fn split(text: &str, pattern: &str) -> Vec<String> {
    let re = regex::Regex::new(pattern).expect("invalid regex");
    re.split(text).map(|s| s.to_string()).collect()
}

/// Return `true` if `pattern` matches anywhere in `text`.
pub fn is_match(text: &str, pattern: &str) -> bool {
    let re = regex::Regex::new(pattern).expect("invalid regex");
    re.is_match(text)
}

/// Extract capture-group tuples for every match of `pattern`.
///
/// Each inner `Vec<String>` contains the captured groups (excluding
/// the full match at group 0).
pub fn extract_groups(text: &str, pattern: &str) -> Vec<Vec<String>> {
    let re = regex::Regex::new(pattern).expect("invalid regex");
    re.captures_iter(text)
        .map(|caps| {
            (1..caps.len())
                .map(|i| caps.get(i).map_or(String::new(), |m| m.as_str().to_string()))
                .collect()
        })
        .collect()
}

// ---------------------------------------------------------------------------
// Pre-built patterns (as &str constants)
// ---------------------------------------------------------------------------

pub const EMAIL: &str = r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}";
pub const URL: &str = r#"https?://[^\s<>"']+"#;
pub const PHONE: &str = r"\+?1?[-.\s]?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}";
pub const DATE: &str = r"\d{4}-\d{2}-\d{2}";
pub const IP_ADDRESS: &str = r"\b(?:\d{1,3}\.){3}\d{1,3}\b";

// ---------------------------------------------------------------------------
// Standalone validators (no regex crate needed)
// ---------------------------------------------------------------------------

/// Check whether `s` looks like a valid email address (simple heuristic).
pub fn is_email(s: &str) -> bool {
    let parts: Vec<&str> = s.split('@').collect();
    if parts.len() != 2 {
        return false;
    }
    let local = parts[0];
    let domain = parts[1];
    !local.is_empty()
        && !domain.is_empty()
        && domain.contains('.')
        && domain.split('.').last().map_or(false, |tld| tld.len() >= 2)
        && local
            .chars()
            .all(|c| c.is_alphanumeric() || "._%+-".contains(c))
        && domain
            .chars()
            .all(|c| c.is_alphanumeric() || ".-".contains(c))
}

/// Check whether `s` looks like a valid IPv4 address.
pub fn is_ipv4(s: &str) -> bool {
    let octets: Vec<&str> = s.split('.').collect();
    octets.len() == 4
        && octets
            .iter()
            .all(|o| o.parse::<u8>().is_ok() && !o.starts_with('0') || *o == "0")
}

/// Find all email addresses in `text` using character scanning.
pub fn find_emails(text: &str) -> Vec<String> {
    text.split_whitespace()
        .filter_map(|word| {
            let trimmed = word.trim_matches(|c: char| !c.is_alphanumeric() && c != '@' && c != '.' && c != '_' && c != '%' && c != '+' && c != '-');
            if is_email(trimmed) {
                Some(trimmed.to_string())
            } else {
                None
            }
        })
        .collect()
}

/// Find all IPv4 addresses in `text`.
pub fn find_ipv4s(text: &str) -> Vec<String> {
    text.split_whitespace()
        .filter_map(|word| {
            let trimmed = word.trim_matches(|c: char| !c.is_ascii_digit() && c != '.');
            if is_ipv4(trimmed) {
                Some(trimmed.to_string())
            } else {
                None
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_email() {
        assert!(is_email("alice@example.com"));
        assert!(is_email("bob.smith+tag@work.org"));
        assert!(!is_email("not-an-email"));
        assert!(!is_email("@missing-local.com"));
        assert!(!is_email("missing-domain@"));
    }

    #[test]
    fn test_is_ipv4() {
        assert!(is_ipv4("192.168.1.1"));
        assert!(is_ipv4("10.0.0.1"));
        assert!(is_ipv4("0.0.0.0"));
        assert!(!is_ipv4("999.999.999.999"));
        assert!(!is_ipv4("1.2.3"));
        assert!(!is_ipv4("not.an.ip.addr"));
    }

    #[test]
    fn test_find_emails() {
        let text = "Mail alice@example.com and bob@work.org please.";
        let found = find_emails(text);
        assert_eq!(found, vec!["alice@example.com", "bob@work.org"]);
    }

    #[test]
    fn test_find_ipv4s() {
        let text = "Servers at 192.168.1.1 and 10.0.0.1 are up.";
        let found = find_ipv4s(text);
        assert_eq!(found, vec!["192.168.1.1", "10.0.0.1"]);
    }

    #[test]
    fn test_find_matches() {
        let text = "abc 123 def 456";
        let matches = find_matches(text, r"\d+");
        assert_eq!(matches, vec!["123", "456"]);
    }

    #[test]
    fn test_find_with_positions() {
        let text = "hello world";
        let matches = find_with_positions(text, r"\w+");
        assert_eq!(matches.len(), 2);
        assert_eq!(matches[0].text, "hello");
        assert_eq!(matches[0].start, 0);
        assert_eq!(matches[0].end, 5);
    }

    #[test]
    fn test_replace() {
        assert_eq!(
            replace("2026-04-16", r"(\d{4})-(\d{2})-(\d{2})", "$3/$2/$1"),
            "16/04/2026"
        );
    }

    #[test]
    fn test_split() {
        let parts = split("a, b, c", r",\s*");
        assert_eq!(parts, vec!["a", "b", "c"]);
    }

    #[test]
    fn test_is_match() {
        assert!(is_match("hello world", r"wo.ld"));
        assert!(!is_match("hello world", r"^\d+$"));
    }

    #[test]
    fn test_extract_groups() {
        let text = "[2026-04-16 08:30] ERROR: disk full";
        let groups = extract_groups(text, r"\[(\d{4}-\d{2}-\d{2}) (\d{2}:\d{2})\] (\w+): (.+)");
        assert_eq!(groups.len(), 1);
        assert_eq!(groups[0], vec!["2026-04-16", "08:30", "ERROR", "disk full"]);
    }
}
